src/cursor/abstract_cursor.ts

```diff
@@ -0,0 +1,710 @@
+    } else {
+      throw new TypeError('Invalid read preference: ' + readPreference);
+    }
+
+    return this;
+  }
+
+  /**
+   * Set a maxTimeMS on the cursor query, allowing for hard timeout limits on queries (Only supported on MongoDB 2.6 or higher)
+   *
+   * @param value - Number of milliseconds to wait before aborting the query.
+   */
+  maxTimeMS(value: number): this {
+    assertUninitialized(this);
+    if (typeof value !== 'number') {
+      throw new TypeError('maxTimeMS must be a number');
+    }
+
+    this[kOptions].maxTimeMS = value;
+    return this;
+  }
+
+  /**
+   * Set the batch size for the cursor.
+   *
+   * @param value - The number of documents to return per batch. See {@link https://docs.mongodb.com/manual/reference/command/find/|find command documentation}.
+   */
+  batchSize(value: number): this {
+    assertUninitialized(this);
+    if (this[kOptions].tailable) {
+      throw new MongoError('Tailable cursors do not support batchSize');
+    }
+
+    if (typeof value !== 'number') {
+      throw new TypeError('batchSize requires an integer');
+    }
+
+    this[kOptions].batchSize = value;
+    return this;
+  }
+
+  /* @internal */
+  abstract _initialize(
+    session: ClientSession | undefined,
+    callback: Callback<ExecutionResult>
+  ): void;
+
+  /* @internal */
+  _getMore(batchSize: number, callback: Callback<Document>): void {
+    const cursorId = this[kId];
+    const cursorNs = this[kNamespace];
+    const server = this[kServer];
+
+    if (cursorId == null) {
+      callback(new MongoError('Unable to iterate cursor with no id'));
+      return;
+    }
+
+    if (server == null) {
+      callback(new MongoError('Unable to iterate cursor without selected server'));
+      return;
+    }
+
+    server.getMore(
+      cursorNs.toString(),
+      cursorId,
+      {
+        ...this[kOptions],
+        session: this[kSession],
+        batchSize
+      },
+      callback
+    );
+  }
+}
+
+function nextDocument(cursor: AbstractCursor): Document | null | undefined {
+  if (cursor[kDocuments] == null || !cursor[kDocuments].length) {
+    return null;
+  }
+
+  const doc = cursor[kDocuments].shift();
+  if (doc) {
+    const transform = cursor[kTransform];
+    if (transform) {
+      return transform(doc);
+    }
+
+    return doc;
+  }
+
+  return null;
+}
+
+function next(cursor: AbstractCursor, callback: Callback<Document | null>): void {
+  const cursorId = cursor[kId];
+  if (cursor.closed) {
+    return callback(undefined, null);
+  }
+
+  if (cursor[kDocuments] && cursor[kDocuments].length) {
+    callback(undefined, nextDocument(cursor));
+    return;
+  }
+
+  if (cursorId == null) {
+    // All cursors must operate within a session, one must be made implicitly if not explicitly provided
+    if (cursor[kSession] == null && cursor[kTopology].hasSessionSupport()) {
+      cursor[kSession] = cursor[kTopology].startSession({ owner: cursor, explicit: true });
+    }
```

Comment:
Do we need to be concerned about what happens if `cursor[kTopology].hasSessionSupport()` is false here?